plugins/ltac2/tac2core.ml

```diff
@@ -1076,247 +962,230 @@ let () = define1 "case" closure begin fun f ->
+  let ty = repr_ext val_free in
+  define "fresh_free_union" (ty @-> ty @-> ret ty) Id.Set.union
 
-let () = define1 "fresh_free_of_ids" (list ident) begin fun ids ->
-  let free = List.fold_right Id.Set.add ids Id.Set.empty in
-  return (Value.of_ext Value.val_free free)
-end
+let () =
+  define "fresh_free_of_ids" (list ident @-> ret (repr_ext val_free)) @@ fun ids ->
+  List.fold_right Id.Set.add ids Id.Set.empty
 
-let () = define1 "fresh_free_of_constr" constr begin fun c ->
+let () =
+  define "fresh_free_of_constr" (constr @-> tac (repr_ext val_free)) @@ fun c ->
   Proofview.tclEVARMAP >>= fun sigma ->
-  let rec fold accu c = match EConstr.kind sigma c with
-  | Constr.Var id -> Id.Set.add id accu
-  | _ -> EConstr.fold sigma fold accu c
+  let rec fold accu c =
+    match EConstr.kind sigma c with
+    | Constr.Var id -> Id.Set.add id accu
+    | _ -> EConstr.fold sigma fold accu c
   in
-  let ans = fold Id.Set.empty c in
-  return (Value.of_ext Value.val_free ans)
-end
+  return (fold Id.Set.empty c)
 
-let () = define2 "fresh_fresh" (repr_ext val_free) ident begin fun avoid id ->
-  let nid = Namegen.next_ident_away_from id (fun id -> Id.Set.mem id avoid) in
-  return (Value.of_ident nid)
-end
+let () =
+  define "fresh_fresh" (repr_ext val_free @-> ident @-> ret ident) @@ fun avoid id ->
+  Namegen.next_ident_away_from id (fun id -> Id.Set.mem id avoid)
 
 (** Env *)
 
-let () = define1 "env_get" (list ident) begin fun ids ->
-  let r = match ids with
+let () =
+  define "env_get" (list ident @-> ret (option reference)) @@ fun ids ->
+  match ids with
   | [] -> None
   | _ :: _ as ids ->
     let (id, path) = List.sep_last ids in
     let path = DirPath.make (List.rev path) in
     let fp = Libnames.make_path path id in
     try Some (Nametab.global_of_path fp) with Not_found -> None
-  in
-  return (Value.of_option Value.of_reference r)
-end
 
-let () = define1 "env_expand" (list ident) begin fun ids ->
-  let r = match ids with
+let () =
+  define "env_expand" (list ident @-> ret (list reference)) @@ fun ids ->
+  match ids with
   | [] -> []
   | _ :: _ as ids ->
     let (id, path) = List.sep_last ids in
     let path = DirPath.make (List.rev path) in
     let qid = Libnames.make_qualid path id in
     Nametab.locate_all qid
-  in
-  return (Value.of_list Value.of_reference r)
-end
 
-let () = define1 "env_path" reference begin fun r ->
+let () =
+  define "env_path" (reference @-> tac (list ident)) @@ fun r ->
   match Nametab.path_of_global r with
   | fp ->
     let (path, id) = Libnames.repr_path fp in
     let path = DirPath.repr path in
-    return (Value.of_list Value.of_ident (List.rev_append path [id]))
+    return (List.rev_append path [id])
   | exception Not_found ->
     throw err_notfound
-end
 
-let () = define1 "env_instantiate" reference begin fun r ->
+let () =
+  define "env_instantiate" (reference @-> tac constr) @@ fun r ->
   Proofview.tclENV >>= fun env ->
   Proofview.tclEVARMAP >>= fun sigma ->
```

Comment:
>For the record, here is an alternative definition I considered,

I guess it would be for cases like this?

Reply:
Yes, that would be one example.